File: javascript/src/screenplay.js
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2024 Ian Thomas

export const ElementType = Object.freeze({
    TITLEENTRY: 'TITLEENTRY',
    HEADING: 'HEADING',
    ACTION: 'ACTION',
    CHARACTER: 'CHARACTER',
    DIALOGUE: 'DIALOGUE',
    PARENTHETICAL: 'PARENTHETICAL',
    LYRIC: "LYRIC",
    TRANSITION: "TRANSITION",
    PAGEBREAK: "PAGEBREAK",
    NOTE: "NOTE",
    BONEYARD: "BONEYARD",
    SECTION: "SECTION",
    SYNOPSIS: "SYNOPSIS"
});

// Base class for all elements
export class Element {
    constructor(type, text) {
        this.type = type;
        this._text = text;
        this.tags = [];
    }

    // This version will not contain any annotations / note markup
    get text() {
        const regex = /\[\[\d+\]\]|\/*\d+\*\//g;
        return this._text.replace(regex, "");
    }

    // Returns with embedded notes (e.g. [[1]] means notes index 1) or boneyards (e.g. /*2*/ means boneyard index 2)
    get textRaw() {
        return this._text;
    }

    appendLine(line) {
        this._text += "\n" + line;
    }

    appendTags(tags) {
        this.tags = this.tags.concat(tags.filter(item => !this.tags.includes(item)));
    }

    // For debugging
    dump() {
        return `${this.type}:"${this._text}"`;
    }
}


export class TitleEntry extends Element {
    constructor(key, text) {
        super(ElementType.TITLEENTRY, text);
        this.key = key;
    }

    // For debugging
    dump() {
        return `${this.type}:"${this.key}":"${this._text}"`
    }
}


export class Action extends Element {
    constructor(text, forced = false) {
        super(ElementType.ACTION, text);
        this.centered = false;
        this.forced = forced;
    }

    // For debugging
    dump() {
        let out = `${this.type}:"${this._text}"`
        if (this.centered)
            out += " (centered)";
        return out;
    }
}


export class SceneHeading extends Element {
    constructor(text, sceneNumber, forced = false) {
        super(ElementType.HEADING, text);
        this.sceneNumber = sceneNumber;
        this.forced = forced;
    }

    // For debugging
    dump() {
        let out = `${this.type}:"${this.text}"`;
        if (this.sceneNumber) {
            out += ` (${this.sceneNumber})`;
        }
        return out;
    }
}


export class Character extends Element {
    constructor(name, extension, dual, forced = false) {
        super(ElementType.CHARACTER, "");
        this.name = name;
        this.extension = extension;
        this.isDualDialogue = dual;
        this.forced = forced;
    }

    // For debugging
    dump() {
        let out = `${this.type}:"${this.name}"`;
        if (this.extension) {
            out += ` "(${this.extension})"`;
        }
        if (this.isDualDialogue)
            out += ` (Dual)`;
        return out;
    }
}


export class Dialogue extends Element {
    constructor(text) {
        super(ElementType.DIALOGUE, text);
    }
}


export class Parenthetical extends Element {
    constructor(text) {
        super(ElementType.PARENTHETICAL, text);
    }
}


export class Lyric extends Element {
    constructor(text) {
        super(ElementType.LYRIC, text);
    }
}


export class Transition extends Element {
    constructor(text, forced = false) {
        super(ElementType.TRANSITION, text);
        this.forced = forced;
    }
}


export class PageBreak extends Element {
    constructor() {
        super(ElementType.PAGEBREAK, "")
    }
}


export class Note extends Element {
    constructor(text) {
        super(ElementType.NOTE, text);
    }
}


export class Boneyard extends Element {
    constructor(text) {
        super(ElementType.BONEYARD, text);
    }
}


export class Section extends Element {
    constructor(level, text) {
        super(ElementType.SECTION, text);
        this.level = level;
    }

    // For debugging
    dump() {
        return `${this.type}:"${this._text}" (${this.level})`;
    }
}

export class Synopsis extends Element {
    constructor(text) {
        super(ElementType.SYNOPSIS, text);
    }
}


// Parsed script
export class Script {

    constructor() {
        this.titleEntries = [];
        this.elements = [];
        this.notes = [];
        this.boneyards = [];
        this.lastChar = null;
    }

    dump() {
        let lines = [];
        for (const entry of this.titleEntries) {
            if (entry.tags.length > 0)
                lines.push(`${entry.dump()} tags:${entry.tags}`);
            else
                lines.push(`${entry.dump()}`);
        }
        for (const element of this.elements) {
            if (element.tags.length > 0)
                lines.push(`${element.dump()} tags:${element.tags}`);
            else
                lines.push(`${element.dump()}`);
        }
        let i = 0;
        for (const note of this.notes) {
            lines.push(`[[${i}]]${note.dump()}`);
            i++;
        }
        i = 0;
        for (const boneyard of this.boneyards) {
            lines.push(`/*${i}*/${boneyard.dump()}`);
            i++;
        }
        return lines.join("\n");
    }

    getLastElem() {
        if (this.elements.length == 0)
            return null;
        return this.elements[this.elements.length - 1];
    }

    addElement(elem, allowMerge = false) {
        let lastElem = this.getLastElem();
        if (elem.type == ElementType.CHARACTER) {
            let newChar = elem.name + (elem.extension ? elem.extension : "");
            if (allowMerge && this.lastChar == newChar)
                return;
            this.lastChar = newChar;
        }

        else if (elem.type == ElementType.DIALOGUE) {
            if (allowMerge && lastElem && lastElem.type == ElementType.DIALOGUE) {
                lastElem._text += "\n" + elem._text;
                return;
            }
        }

        else if (elem.type == ElementType.PARENTHETICAL) {
            //
        }

        else {
            this.lastChar = null;
        }

        if (elem.type == ElementType.ACTION) {
            if (allowMerge && lastElem && lastElem.type == ElementType.ACTION) {
                lastElem._text += "\n" + elem._text;
                return;
            }
        }

        this.elements.push(elem);
    }
}

File: cpp/src/screenplay.cpp
// This file is part of an MIT-licensed project: see LICENSE file or README.md
// for details. Copyright (c) 2024 Ian Thomas

#include "screenplay_tools/screenplay.h"
#include "screenplay_tools/utils.h"
#include <regex>
#include <unordered_map>

namespace ScreenplayTools {

// Utility function to convert enums to strings
std::string elementTypeToString(ElementType type) {
  static const std::unordered_map<ElementType, std::string> elementStrings = {
      {ElementType::TITLEENTRY, "TITLEENTRY"},
      {ElementType::HEADING, "HEADING"},
      {ElementType::ACTION, "ACTION"},
      {ElementType::CHARACTER, "CHARACTER"},
      {ElementType::DIALOGUE, "DIALOGUE"},
      {ElementType::PARENTHETICAL, "PARENTHETICAL"},
      {ElementType::LYRIC, "LYRIC"},
      {ElementType::TRANSITION, "TRANSITION"},
      {ElementType::PAGEBREAK, "PAGEBREAK"},
      {ElementType::NOTE, "NOTE"},
      {ElementType::BONEYARD, "BONEYARD"},
      {ElementType::SECTION, "SECTION"},
      {ElementType::SYNOPSIS, "SYNOPSIS"}};

  auto it = elementStrings.find(type);
  return it != elementStrings.end() ? it->second : "UNKNOWN";
}

// Element class implementations
std::string Element::dump() const {
  return elementTypeToString(_type) + ":\"" + _textRaw + "\"";
}

void Element::_updateText() {
  const std::regex regex(R"(\[\[\d+\]\]|\/*\d+\*\/)");
  _textClean = std::regex_replace(_textRaw, regex, "");
}

// TitleEntry
std::string TitleEntry::dump() const {
  return getTypeAsString() + ":\"" + _key + "\":\"" + getTextRaw() + "\"";
}

// Action
std::string Action::dump() const {
  std::string output = getTypeAsString() + ":\"" + getTextRaw() + "\"";
  if (_centered)
    output += " (centered)";
  return output;
}

// SceneHeading
std::string SceneHeading::dump() const {
  std::string output = getTypeAsString() + ":\"" + getText() + "\"";
  if (_sceneNumber.has_value()) {
    output += " (" + _sceneNumber.value() + ")";
  }
  return output;
}

// Character
std::string Character::dump() const {
  std::string output = getTypeAsString() + ":\"" + _name + "\"";
  if (_extension.has_value() && !_extension->empty()) {
    output += " \"(" + *_extension + ")\"";
  }
  if (_isDualDialogue) {
    output += " (Dual)";
  }
  return output;
}

// Section
std::string Section::dump() const {
  return getTypeAsString() + ":\"" + getTextRaw() + "\" (" +
         std::to_string(_level) + ")";
}

// Script
std::string Script::dump() const {

  std::vector<std::string> lines;

  for (const auto &titleEntry : _titleEntries) {
    if (titleEntry->getTags().size() > 0)
      lines.push_back(titleEntry->dump() +
                      " tags:" + join(titleEntry->getTags(), ","));
    else
      lines.push_back(titleEntry->dump());
  }

  for (const auto &element : _elements) {
    if (element->getTags().size() > 0)
      lines.push_back(element->dump() +
                      " tags:" + join(element->getTags(), ","));
    else
      lines.push_back(element->dump());
  }

  for (int i = 0; i < _notes.size(); i++) {
    lines.push_back("[[" + std::to_string(i) + "]]" + _notes[i]->dump());
  }
  for (int i = 0; i < _boneyards.size(); i++) {
    lines.push_back("/*" + std::to_string(i) + "*/" + _boneyards[i]->dump());
  }
  return join(lines, "\n");
}

void Script::addElement(const std::shared_ptr<Element> &element,
                        bool allowMerge) {

  std::shared_ptr<Element> lastElem = getLastElement();

  if (element->getType() == ElementType::CHARACTER) {
    std::shared_ptr<Character> charElem =
        std::dynamic_pointer_cast<Character>(element);
    std::string newChar =
        charElem->getName() + (charElem->getExtension().has_value()
                                   ? charElem->getExtension().value()
                                   : "");
    if (allowMerge && _lastChar == newChar)
      return;
    _lastChar = newChar;
  }

  else if (element->getType() == ElementType::DIALOGUE) {
    if (allowMerge && lastElem &&
        lastElem->getType() == ElementType::DIALOGUE) {
      lastElem->appendLine(element->getTextRaw());
      return;
    }
  }

  else if (element->getType() == ElementType::PARENTHETICAL) {
    //
  }

  else {
    _lastChar = "";
  }

  if (element->getType() == ElementType::ACTION) {
    if (allowMerge && lastElem && lastElem->getType() == ElementType::ACTION) {
      lastElem->appendLine(element->getTextRaw());
      return;
    }
  }

  _elements.push_back(element);
}

} // namespace ScreenplayTools
